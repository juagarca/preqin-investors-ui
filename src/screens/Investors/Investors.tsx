import { useQuery } from "@tanstack/react-query";
import { useNavigate } from "react-router-dom";

import { InvestorsTable } from "../../components";

import { fetchInvestors } from "../../utils/api";
import ROUTES from "../../routes";

const Investors = () => {
  const navigate = useNavigate();

  const {
    data: investors,
    isLoading,
    error,
  } = useQuery({
    queryKey: ["investors"],
    queryFn: () => fetchInvestors(),
  });

  if (isLoading) return <div>Loading...</div>;
  if (error || !investors) return <div>Error!</div>;

  const handleRowClick = (firmId: number, firmName: string) => {
    navigate(`${ROUTES.investors}/${firmId}`, {
      state: { firmName: firmName },
    });
  };

  return (
    <div>
      <h1>Investors</h1>
      {investors.length > 1 ? (
        <InvestorsTable investors={investors} handleRowClick={handleRowClick} />
      ) : (
        <p>No investors found</p>
      )}
    </div>
  );
};

export default Investors;
