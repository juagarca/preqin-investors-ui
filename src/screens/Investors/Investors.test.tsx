import { fireEvent, render, screen } from "@testing-library/react";
import { useQuery } from "@tanstack/react-query";
import { MemoryRouter, Route, Routes } from "react-router-dom";

import Investors from "./Investors";

import investors from "../../utils/tests/mocks/investors.json";
import ROUTES from "../../routes";

const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));
jest.mock("@tanstack/react-query");

describe("Investors component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = () => {
    return render(
      <MemoryRouter initialEntries={[ROUTES.investors]}>
        <Routes>
          <Route path={ROUTES.investors} element={<Investors />} />
        </Routes>
      </MemoryRouter>
    );
  };

  it("matches the snapshot", () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: investors,
      isLoading: false,
      error: false,
    });

    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders successfully when investors data is found", () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: investors,
      isLoading: false,
      error: false,
    });

    renderComponent();

    expect(screen.getByText("Investors")).toBeInTheDocument();

    investors.forEach((investor) => {
      expect(screen.getByText(investor.firm_name)).toBeInTheDocument();
    });
  });

  it("navigates to the correct route when clicking on a row", () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: investors,
      isLoading: false,
      error: false,
    });

    renderComponent();

    const rows = screen.getAllByRole("row");
    fireEvent.click(rows[1]);

    expect(mockNavigate).toHaveBeenCalledWith(
      `${ROUTES.investors}/${investors[0].firm_id}`,
      {
        state: { firmName: investors[0].firm_name },
      }
    );
  });

  it("renders message when no investors data found", () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: [],
      isLoading: false,
      error: false,
    });

    renderComponent();

    expect(screen.getByText("No investors found")).toBeInTheDocument();
  });

  it('renders loading when fetching"', () => {
    (useQuery as jest.Mock).mockReturnValue({
      isLoading: true,
    });

    renderComponent();

    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("renders error when a fetching error occurs", () => {
    (useQuery as jest.Mock).mockReturnValue({
      error: true,
    });

    renderComponent();

    expect(screen.getByText("Error!")).toBeInTheDocument();
  });
});
