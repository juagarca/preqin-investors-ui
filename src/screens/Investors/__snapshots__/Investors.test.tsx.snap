// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investors component matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Investors
    </h1>
    <table
      class="inverstors-table"
    >
      <thead>
        <tr>
          <th>
            Firm ID
          </th>
          <th>
            Firm Name
          </th>
          <th>
            Type
          </th>
          <th>
            Date added
          </th>
          <th>
            Address
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          dataset-firm-id="2670"
          dataset-firm-name="Mjd Jedi fund"
        >
          <td>
            2670
          </td>
          <td>
            Mjd Jedi fund
          </td>
          <td>
            Bank
          </td>
          <td>
            8th June 2010
          </td>
          <td>
            29 Nathan Road, Hong Kong
          </td>
        </tr>
        <tr
          dataset-firm-id="2792"
          dataset-firm-name="Ibx Skywalker ltd"
        >
          <td>
            2792
          </td>
          <td>
            Ibx Skywalker ltd
          </td>
          <td>
            Asset manager
          </td>
          <td>
            21st July 1997
          </td>
          <td>
            19 Fifth Avenue, New York
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
